### PR TITLE
sql: make column backfill batch size a setting

### PR DESCRIPTION
**Pull request by dt**
Release note: none.

### CODE REVIEW

**Review by adityamaru — APPROVED** [review]

File: pkg/sql/backfill.go [hunk 1]
```diff
@@ -87,6 +80,16 @@ var indexBackfillBatchSize = settings.RegisterIntSetting(
 	settings.NonNegativeInt, /* validateFn */
 )
 
+// indexBackfillBatchSize is the maximum number of rows we construct index
```
Comment (adityamaru) [inline pkg/sql/backfill.go:83]: nit: comment needs updating

**Comment by ajwerner** [comment]
Relates to #63634.

**Comment by dt** [comment]
TFTR

bors r+

**Comment by daniel-crlabs** [comment]
A customer is impacted by this. Do we know which cockroachdb release version this will be available? Client is asking, they are running on 20.2.7. Please let us know, thanks.

**Comment by dt** [comment]
They've already tried putting the new column in its own family?  given the limit was a mere 200 rows even before this knob, it must be a pretty large value to be hitting it.

I think the next 20.2 release is May 17, 2021. Could recommend a 20.2 nightly in the meantime.

**Comment by daniel-crlabs** [comment]
Thanks for the response, they were adding a brand new column and not as part of a family.
This is the original table ddl:

```
show create aggregation_history;
table_name | create_statement
-----------------------------+-----------------------------------------------------------------------------------------------------------------------------------
aggregation_history | CREATE TABLE public.aggregation (
| faux_id INT8 NOT NULL DEFAULT unique_rowid(),
| aggregation_id UUID NOT NULL,
| entity_type STRING NOT NULL,
| entity_id STRING NOT NULL,
| created_at TIMESTAMP NOT NULL DEFAULT now():::TIMESTAMP,
| data JSONB NULL,
| version DECIMAL NULL DEFAULT 0:::DECIMAL,
| change_type STRING NULL,
| change_reference STRING NULL,
| CONSTRAINT "primary" PRIMARY KEY (aggregation_id ASC, entity_id ASC, entity_type ASC, created_at ASC, faux_id ASC),
| FAMILY "primary" (faux_id, aggregation_id, entity_type, entity_id, created_at, data, version, change_type, change_reference)
| );
```

They just tried to do the following and got the error:

```
ALTER TABLE aggregation_history
ADD COLUMN ehid string as ( md5(concat(entity_id, entity_type)) ) STORED;
ERROR: command is too large: 268046651 bytes (max: 67108864)
```

**Comment by dt** [comment]
huh, it sounds likely that one of the other columns here is already very large  -- a 268mb write for a 200 row batch sounds like they have ~1mb rows here?

putting the new column in its own family would help, or using a nightly with the setting to do fewer rows at once.

**Comment by daniel-crlabs** [comment]
I'm not sure that I understand exactly. Is there a way to do an alter statement as the above but telling it to alter the table but only 50 rows at a time? I'm not very familiar with the inner workings yet of crdb, but I know others create a copy of the table in the background (like a ghost table per se) with the new columns, then load the data from current to new table, then atomically swap the tables. Is this how crdb does as well?

**Comment by dt** [comment]
that is how some future version of cockroach will do column additions and removals, since that's how we do e.g. primary key changes, but today, we instead iterate through all the rows and update them to add or remove the new/old columns. We do that in batches. Prior to this change, we did them in batches of 200 rows at a time. After this change it is a cluster-wide config setting. But if the current code, at 200 rows at a time, is hitting the error `command is too large: 268046651 bytes`, that means a batch of 200 rows produced a 270mb command, so we must be talking about some _big_ rows. 

Often, if we see one very large value, we suggest putting it in a separate column family from the rest of the columns. That way edits to those columns won't need to rewrite the big column, since when any column in a family is edited, the whole family is rewritten.

We _could_ put the new column being added in its own family, so that we wouldn't rewrite all the rows for the giant column to add it, but that is a little weird, since then we'd have a tiny column in its own family. At this point the best bet is probably wait for 20.2.8 or use a nightly in the meantime and set the batch size to like 30 or 50 to get this change done, then maybe after that look at moving whatever the giant 1mb/row column is to its own family.

**Pull request merged**